R/io.R

```diff
@@ -190,3 +220,62 @@ mask_raster_to_polygon <- function(raster_object, polygon) {
 
   return(n_lcs)
 }
+
+#' reclassify_raster
+#' @param crop_cover_levels levels of the crop cover raster file.
+#' @details Read internal data file that specifies the GCAM classification for certain crop types. Reclassify CDL based on GCAM.
+#' @import dplyr group_indices left_join filter rename
+#' @importFrom car recode
+#' @author Kristian Nelson (kristian.nelson@pnnl.gov)
+#' @export
+reclassify_raster <- function(crop_cover_levels){
+  # Load in the GCAM classes CSV.
+  gcam_csv <- get_crop_mapping()
+
+  # Extract only the GCAM Class column
+  gcam_csv %>%
+    dplyr::select(GCAM_commodity) %>%
+    filter(!is.na(GCAM_commodity)) %>%
+    mutate(GCAM_ID = dplyr::group_indices(., GCAM_commodity)) %>%
+    unique() ->
+    gcam_classes
+
+  # Obtain the attribute table for the raster.
+  crop_cover_levels %>%
+    filter(Class_Names != "") %>%
+    dplyr::select(ID, CDL_Class = Class_Names) %>%
+    mutate(CDL_Class = as.character(CDL_Class))->
+    cdl_classes
+
+  # Assign the GCAM IDs to matching CDL Class.
+        # All Double Crops are under MiscCrop
+        # CDL crops were assigned to where they fit best if there was no overlap with GCAM classes.
+        # Fallow/Idle Cropland assigned to MiscCrop because it was not present in GCAM classes.
+        # Any crop that was potentially used for oil is assigned to OilCrop. Example: Camelina
+        # Cantelopes is not present in GCAM classes and is counted as MiscCrop because Watermelon is MiscCrop. Both are melons.
+        # Any grain that was not wheat was assigned to OtherGrain
+        # 1 = Corn; 2 = FiberCrop; 3 = FodderGrass; 4 = FodderHerb; 5 = MiscCrop; 6 = OilCrop;
+        # 7 = OtherGrain; 8 = PalmFruit; 9 = Rice; 10 = Root_Tuber; 11 = SugarCrop; 12 = Wheat
+  cdl_classes$GCAM_ID <- car::recode(cdl_classes$ID, "c(1,12,13) = 1;
```

Comment:
Is there a way that we could read in a YAML or CSV file that maps GCAM land cover ids to CDL land cover ids?  To build this reclass string?  This would allow us to modify these externally without someone needing to interact with (screw up) the code to so.  See attached files.  
[lc_mapping.zip](https://github.com/IMMM-SFA/teleconnect/files/3704178/lc_mapping.zip)

Reply:
@crvernon agree that this can be external table, and in fact Kristian had already done this matching work manually before writing the above code. I challenged him to see if he could produce an interpretable chunk of R code that would allow others to understand the decisions in matching CDL to GCAM. @KristianNelson great job, but we do need to deal with those numbers, which are too difficult to interpret. I suggest that we can merge this as is (to at least have the updated crop counting functionality added), and raise a new issue to come back and address the matching by either (a) tidying away into an external mapping file, or, if possible, (b) match the columns algorithmically, catching remaining cases with more legible code (`dplyr::case_when()`)